src/main/java/org/jenkinsci/plugins/workflow/cps/DSL.java

```diff
@@ -260,7 +260,12 @@ protected Object invokeStep(StepDescriptor d, String name, Object args) {
             }
             d.checkContextAvailability(context);
             Thread.currentThread().setContextClassLoader(CpsVmExecutorService.ORIGINAL_CONTEXT_CLASS_LOADER.get());
-            s = d.newInstance(ps.namedArgs);
+            if (Util.isOverridden(StepDescriptor.class, d.getClass(), "newInstance", Map.class)) {
+                s = d.newInstance(ps.namedArgs);
+            } else {
+                DescribableModel<?> stepModel = DescribableModel.of(d.clazz);
+                s = (Step) stepModel.instantiate(ps.namedArgs, handle.getListener());
```

Comment:
The cast is unnecessary if you do not throw away the type restriction `? extends Step` which `of` correctly produced:

```java
s = DescribableModel.of(d.clazz).instantiate(ps.namedArgs, handle.getListener());
```